### ISSUE TITLE
Create peergos:// URL scheme/standard

### ISSUE THREAD

**Issue by RangerMauve** [Peergos/Peergos#963]
Hey folks, 

I've been talking to @ianopolous on Matrix about Peergos and have been more curious about  using it in a p2p web browsers.

For context, I'm the dev of the [Agregore Browser](https://agregore.mauve.moe/) and have been working on mixing p2p protocols on the web. Here's a [talk I recently did that talks about it](https://www.youtube.com/watch?v=W5vNnFTjBI8).

I was thinking it'd be useful to standardize a URL scheme / protocol handler for peergos so that we could send links around and load them in browsers that support the scheme.

Ian mentioned that data can be linked to using a combination of `owner public key, writer public key, 32 byte label, 32 byte symmetric key`.

Generally, browsers and most tools recognize URLs that look like `scheme://username:passowrd@hostname/subpath` which is what I've been trying to fit most things into.

Are there clear ways that we could think of peergos links in the same format? Also is this something folks would be interested in collaborating on in general? I'd love to get something together into Agregore and maybe eventually Agregore Mobile. 😁

**Comment by jonaswre** [comment]
Hey,

Would you run a peergos instance in the browser? 

As I unterstand brave does this with ipfs:// but the reason for that is it directly uses ipfs to load the files. 
Not going through the gateway. 
But from my unterstanding going from https:// to peergos:// will make sense if this actually allows you to get rid of https (SSL Trust Chain).

@ianopolous accounts belong to the instance right? Do you always need to specify on which instance you login to?

**Comment by ianopolous** [comment]
@jonaswre Yep, you can log in through any instance, and writes are proxied over p2p streams to your home instance.

**Comment by RangerMauve** [comment]
For Agregore I'd be bundling in a peergos instance and transparently routing between the browser's protocol handlers and it. 😁

I'm not sure if all of the functionality would make sense to expose over protocol handlers though. 🤷

**Comment by RangerMauve** [comment]
Regarding brave, under the hood it actually redirects all `ipfs://` URLs to it's built in (or external) gateway.

Agregore also bundles a go-ipfs node, but it doesn't expose it to users and uses it for stuff like PUT/POST and IPNS.

**Comment by ianopolous** [comment]
It's actually difficult to fit caps into this. For the web the hostname is both the server to send queries to (meaningless for us), and also the security isolation boundary. 
Maybe something like peergos://owner.writer.label.bat/subpath#readkey,writekey
That's going to be quite a long host though - more than 128 bytes worth before stringifying.

**Comment by RangerMauve** [comment]
Are there any properties that could be omitted if we assume that the server is running locally? E.g. could we put a bit more trust into the server used for protocol handlers in order to take a bit more out of the URL? Or is that totally counter to the design? 😅

Just looking at the [immutable](https://book.peergos.org/architecture/immutable.html) and [muttable](https://book.peergos.org/architecture/mutable.html) APIs, am I reading it correctly that something like `peergos://owner+writer/subpath` would work for mutable URLs, and `peergos://hash/?bat=BatWithId` would be work for immutable URLs?

Just to confirm, are owner/writer/label/bat all necessary for identifying data? Or could any of that be moved into the `username` or `password` portion of a URL? It seems like readkey and write key would also be obvious candidates for putting into the `username` and `password` slots.

For what its workth, 128 bytes isn't a hard limit in most of the URL implementations I've worked with. E.g. if you run the following code in a JS console, it'll gladly take a hostname with 129 characters.

```JavaScript
new URL(`http://username:password@${new Array(129).fill('A').join('')}/subpath?foo=bar`)
```

**Comment by ianopolous** [comment]
The servers in Peergos are trustless. Everything happens client side normally. 

To retrieve a file/dir from a cap the following happens:
Look up server's nodeID from PKI lookup of owner => nodeX
Proxy mutable get call to nodeX => pointer target (a signed pair of cids)
verify signature of cid pair and take the 2nd (current) cid
Do a champ lookup from that cid using the BAT as permission to retrieve the metadata block.
decrypt the metadata block with the read key.

Then to get the data for the file or child links for a dir
get any linked blocks for this chunk using BATs from the metadata block. 
Retrieve all those blocks, (Max 5 at 1 MiB each) concat and decrypt with a key from the decrypted metadata.
Repeat for any subsequent blocks

The server isn't even aware which different 5MiB chunks belong to the same file.
